src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,320 @@
+#include "common/syncjournaldb.h"
+#include "common/syncjournalfilerecord.h"
+#include "common/utility.h"
+#include "filesystem.h"
+#include "propagatorjobs.h"
+#include "syncengine.h"
+#include "propagateremotemove.h"
+#include "propagateremotedelete.h"
+#include "common/asserts.h"
+#include "configfile.h"
+
+#include <QNetworkAccessManager>
+#include <QFileInfo>
+#include <QDir>
+#include <cmath>
+#include <cstring>
+
+#ifdef Q_OS_UNIX
+#include <unistd.h>
+#include <arpa/inet.h>
+#endif
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcZsyncSeed, "sync.propagate.zsync.seed", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGenerate, "sync.propagate.zsync.generate", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGet, "sync.networkjob.zsync.get", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncPut, "sync.networkjob.zsync.put", QtInfoMsg)
+
+bool isZsyncPropagationEnabled(OwncloudPropagator *propagator, SyncFileItemPtr item)
+{
+    ConfigFile cfg;
+    if (!propagator->account()->capabilities().zsyncSupported()) {
+        qCInfo(lcPropagator) << "[zsync disabled] Lack of server support.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (item->_remotePerm.hasPermission(RemotePermissions::IsMounted) ||
+        item->_remotePerm.hasPermission(RemotePermissions::IsMountedSub)) {
+        qCInfo(lcPropagator) << "[zsync disabled] External storage not supported.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (!(cfg.deltaSyncEnabled() ||
+          propagator->_deltaSyncEnabled.fetchAndAddAcquire(0))) {
+        qCInfo(lcPropagator) << "[zsync disabled] Client configuration option.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (item->_size < cfg.deltaSyncMinFileSize() ||
+        item->_size < propagator->_deltaSyncMinFileSize.fetchAndAddAcquire(0)) {
+        qCInfo(lcPropagator) << "[zsync disabled] File size is smaller than minimum.";
+        item->_isZsync = false;
+        return false;
+    }
+
+    item->_isZsync = true;
+    return true;
+}
+
+void ZsyncSeedJob::seed()
+{
+    // Create a temporary file to use with zsync_begin()
+    QTemporaryFile zsyncControlFile;
+    zsyncControlFile.open();
+    zsyncControlFile.write(_zsyncData.constData(), _zsyncData.size());
+    zsyncControlFile.flush();
+
+    int fileHandle = zsyncControlFile.handle();
+    FILE* f = fdopen(dup(fileHandle), "r");
+    setvbuf(f, NULL, _IONBF, 0);
+    rewind(f);
+
+    struct zsync_state *zs;
+    if ((zs = zsync_begin(f)) == NULL) {
+        QString errorString = tr("Unable to parse zsync file.");
+        emit failedSignal(errorString);
+        return;
+    }
+
+    zsyncControlFile.close();
+    fclose(f);
+
+    std::string tmp = _zsyncFilePath.toStdString();
+    const char * fname = tmp.c_str();
+    {
+        /* Simple uncompressed file - open it */
+        FILE *f = fopen(fname, "r");
+        if (!f) {
+            QString errorString = tr("Unable to open file.");
+            emit failedSignal(errorString);
```

Comment:
missing `zsync_end`

Reply:
That's done by the `failedSignal` handler, reasoning is the same is done by `finishedSignal` handler, keeping logic similar.

Reply:
You're right, sorry I got confused there because of `_zs`, will destroy here.